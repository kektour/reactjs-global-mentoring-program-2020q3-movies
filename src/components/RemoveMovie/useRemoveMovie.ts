import { removeMovie as rM } from '../../services/movies';

type UseRemoveMovie = {
  removeMovie: (id: number) => Promise<void>;
};

export const useRemoveMovie = (): UseRemoveMovie => {
  const removeMovie = (id: number) => rM(id);
  return {
    removeMovie,
  };
};
