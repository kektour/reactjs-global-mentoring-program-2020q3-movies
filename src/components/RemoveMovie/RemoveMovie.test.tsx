import React from 'react';
import ReactDOM from 'react-dom';
import { act, render, wait } from '@testing-library/react';
import { RemoveMovie } from './RemoveMovie';
import * as hooks from './useRemoveMovie';
import userEvent from '@testing-library/user-event';

describe('RemoveMovie', () => {
  beforeAll(() => {
    ReactDOM.createPortal = jest.fn((e) => e) as any;

    const body = document.createElement('body');
    const div = document.createElement('div');
    const spyQuerySelector = jest.fn((selector: string) => {
      if (selector === '#root') {
        return div;
      } else if (selector === 'body') {
        return body;
      }
    });
    Object.defineProperty(global.document, 'querySelector', { value: spyQuerySelector });
  });

  it('Should render component correctly', () => {
    const { asFragment } = render(<RemoveMovie id={1} open={true} onClose={() => null} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should call API on button click', async () => {
    const movieId = 1;
    const removeMovieMock = jest.fn().mockImplementation(() => Promise.resolve());
    const handleCloseMock = jest.fn();
    const useRemoveMovieMock = jest.spyOn(hooks, 'useRemoveMovie').mockImplementation(() => ({ removeMovie: removeMovieMock }));

    const { container } = render(<RemoveMovie id={movieId} open={true} onClose={handleCloseMock} />);

    act(() => userEvent.click(container.querySelector('p + button') as Element));

    await wait(() => {
      expect(removeMovieMock).toHaveBeenCalledWith(movieId);
    });
    expect(handleCloseMock).toHaveBeenCalled();

    useRemoveMovieMock.mockRestore();
  });
});
