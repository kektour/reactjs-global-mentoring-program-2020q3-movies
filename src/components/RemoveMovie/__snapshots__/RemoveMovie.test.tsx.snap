// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoveMovie Should render component correctly 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <div
      class="root header"
    >
      <div
        class="container"
      >
        <div
          class="header"
        >
          <p
            class="root"
          >
            <span>
              netflix
            </span>
            roulette
          </p>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="content"
      >
        <button
          class="crossButton"
        />
        <h2
          class="contentHeader"
        >
          Delete Movie
        </h2>
        <p
          class="contentBody"
        >
          Are you sure you want to delete this movie?
        </p>
        <button
          class="buttonConfirm"
        >
          Confirm
        </button>
      </div>
    </div>
    <div
      class="root footerRoot"
    >
      <p
        class="root logoRoot"
      >
        <span>
          netflix
        </span>
        roulette
      </p>
    </div>
  </div>
</DocumentFragment>
`;
